### ISSUE TITLE
[FEATURE]: Automatically merge objects after state change functions just like zustand does

### ISSUE THREAD

**Issue by Evertt** [state-adapt/state-adapt#92]
In the spirit of removing as much boilerplate as possible, I think zustand made the right choice to merge the object that is returned from a state change function with the original state.

Because that changes a function like this:
```ts
{
  incrementUserAge: state => ({ ...state, age: state.age + 1 }),
}
```

Into this:
```ts
{
  incrementUserAge: ({ age }) => ({ age: age + 1 }),
}
```

Of course I understand that the consequence is that before you could choose to exclude certain keys from the newly returned object, in order to remove them. However:

1. That was a much more rare need than the need to only mutate some properties.
2. That same feature can be offered by deleting keys from objects if the developer returned an object with a key set to undefined.

So, removing a key would essentially become this:

```ts
{
  removeDraft: () => ({ draft: undefined })
}
```

What do you think? I think it would be nice to have this and save a considerable amount of repeated boiler plate.

And of course when the state is not an object, but some kind of primitive (i.e. number, boolean, etc), then this does not apply obviously.

**Comment by mfp22** [comment]
This is the exact reason [`joinAdapters`](https://state-adapt.github.io/docs/core#joinadapters) was created. I pretty much never manage object state without it. 

State can be a primitive type, so it could be very weird trying to figure when to apply this and when not to.

I strongly, strongly prefer solving issues directly rather than "automatically" fixing them with special code. If StateAdapt didn't have `joinAdapters`, this feature request would cause me to invent it now rather than being "smart" and doing sometimes unexpected things.

**Comment by Evertt** [comment]
> This is the exact reason [`joinAdapters`](https://state-adapt.github.io/docs/core#joinadapters) was created. I pretty much never manage object state without it.

Really? So for any object state, you always first define a separate adapter for every single property on that object? Doesn't that make some code much more verbose than it needs to be? I understand how creating and composing small adapters is probably a better pattern in general, for reusability etc. But isn't that pattern overkill / over-engineering in some cases?

> State can be a primitive type, so it could be very weird trying to figure when to apply this and when not to.

How would that be weird? It's pretty straightforward if you ask me. [See a demo here.](https://www.typescriptlang.org/play?target=99#code/MYewdgzgLgBADgJwJYFslSQNwKYBUCec2EMAvDANoBEYArgDb1UA0MVtYAJtgGZJjZOLNgCMQIetgCGYYTVooR2BHJFIA5vyhzoyMOp35FEqgF0YUkqEhQAULaiFsMAArI0GHASIBlKHvUAGXRlKXoSckciEB54d3QsPCcICjpFZVN7KOdcEDdUBK8nAB5cAD4yGFwYbAAPKGwuEipdfgMYAH4YVv0YAC5bGCHhkdGx0eq6hqa2NKUVTpg55X7B8fWNqpr6xs5mtU0wbUWDrVXNi4nt6b3RcUkZKhP76TBzy4-JnZmWozEmRYQP4Sd4fC5fG7NDjcPgCISLaG8fiCUFgjYQ3bNOiMJ5dbH0fpLbA4BBZJyueKeZzkXL5DyJbzYYp0wpJXz+NrBBoIMIQMr2HgcYAYcAwJAQFlUgAUUAAFm0+jAOABrMAgADuYAAlIq5W0xSRJYkYABvNYIbBQWgIN563qkB1LBgEgA+LpgUsQBSpjJIlhgFqknHA9Hw3Q5+gopi1ADp+MB6LRuBAZU4YjA7eotbYAL72AD0+ZgAFUIM4pDBrJwEuAwhnyVAQGKwDwVnLnBarTb60QYCJLCjRe3m3BaLBsg5yQBJCAAeREACtsMLShUaddMRTvcaujxec5dQhaNgANwCoUit4QWUgBicACyynUTPKMvl+kVuB1MBn86XK-KU01kLGAfHjHInB8YBkDgWBgBkNVYG4bk0AEDNZQ7S1rVtcl+zLTgYFFI8jlQZxMDCY8IGYYCiz1EhiIwFBnGADDgGVGBg2IJYQFgKQeFbYV0OYkAUDgJBJAAWkY5xsmbVsLTAYBsBjGiqllWgqJgdVnAEFFG0rSxxwwgMsO7WT9JQKQoBYoTK3AatLzrWS6iIYUrKQcAVKGTtsI9ABCcUjRwN82i1LULBIX9F2XKBVzPPNbGsaAxQI8h8VYDTlDAKQmMqKgACkQFlN4ABEQGwYQpGfSoAGYAFZ7CS2Br1vegHyfbAADlnUqFq70fBBnylJBOGzEDhgAPQ6RrwGSvq2oG58-ACXqb36jqpUym0cuwMai0m6bEtm5q1oWjruvSBBVta9rBuwKUqt22xxqGKaZpsbpTtu59iy4JE4Wu9a7s2v7YUEPaRjeo6Pvm77sD-GLAbO4GTRSjKy22pjWEemAcwhg73ogCRlPoEB1ClM0hlhxauoUeZqKpr6aeWtoGc+m6ad+mFkU4NnqfO50+aZjqEeFai8aAA)

> I strongly, strongly prefer solving issues directly rather than "automatically" fixing them with special code. If StateAdapt didn't have `joinAdapters`, this feature request would cause me to invent it now rather than being "smart" and doing sometimes unexpected things.

Again I agree that this may be the right approach in many (if not most) cases. But if I remember correctly, one of your goals was also to make StateAdapt require as little boilerplate as possible, especially for the simplest use-cases. That's why you allow someone to create a `const nameStore = adapt("Bob")` without an adapter even, because in the simplest case `adapt` could and should just provide its own most minimalistic adapter with state change functions such as `set()` and `reset()`.

So, allowing a developer to write a userStore like this:

```ts
const userStore = adapt({ name: "John Doe", age: 35 }, {
  incrementAge: ({ age }) => ({ age: age + 1 }),
})
```

Just serves the same goal of reducing as much boilerplate as possible, no?

**Comment by mfp22** [comment]
(Just read the first half of this if you want. 2nd half is me brainstorming)

Yeah but there are some steps in syntax that not even I am obsessive enough to try to eliminate. Minimalism shouldn't get in the way of readability. Like, wanting to write objects like `{ name: 'John Doe', age: number }` inline without defining the type is just delaying the inevitable, and making it harder to read in the meantime. It feels like what I did as a junior developer. Most devs would just give a name to the thing they're dealing with:

```typescript
interface User {
  name: string;
  age: number;
}

const initialUser: User = {
  name: 'John Doe',
  age: 35,
};
```

And I know that a simple state change could be easy to express as either returning a partial or a complete state, and use conditional types to do that. But that complexity would then cascade _everywhere_ throughout the codebase. I'm just trying to think of how to apply that in [`joinAdapters`](https://github.com/state-adapt/state-adapt/blob/main/libs/core/src/lib/adapters/join-adapters.types.ts) and the other places that would have to become aware that the return type isn't always just `State`. I'd rather take NGXS' approach of providing a utility like `incrementAge: ({ age }) => patch({ age: age + 1 })`.

But in reality, in the long run, I value reusability and composability more. I want people to think about building up from primitives. In the long run, that's going to result in more minimal code. 

So minimalism isn't the absolute goal... The priorities of StateAdapt are these:

1. Reactivity
2. Reusability
3. Minimalism

This was on the docs homepage originally. It sort of doesn't matter to the everyday developer though, so I changed the homepage. 

Reactivity is by far the highest priority. So I actually prefer the unidirectionality of NgRx/Redux over most state management solutions. I want things to be sane, even if it's visually terrible. 

But other than limited reactivity, reusability was the main frustration I had with NgRx/Redux. I could develop all these state management patterns happily, then the designer could decide, hey, actually we're going to have _2 of these_ on the page together, so now... restructure your state management to account for that. That happened to me twice at different jobs in ways that caused weeks of work.

Basically, I want the actual state management code to be minimized. I wish 90% of it was already written. So I hate the idea of making it easy to re-invent the wheel instead of importing what has already been done 1,000,000 times. Maybe I need to improve `joinAdapters` instead, or integrate it more.

```typescript
import { numberAdapter } from '@state-adapt/core/adapters';

interface User {
  name: string;
  age: number;
}

const initialUser: User = {
  name: 'John Doe',
  age: 35,
};

const userStore = adapt(initialUser, joinAdapters<User, 'name'>()({
  age: numberAdapter
})();
```

Here are some complaints I have with `joinAdapters`:

1. To get it to type-check the adapters you're passing in, it has to require either an adapter for every property, or to specify when a property will not get an adapter (like `<User, 'name'>` in that example)
2. It seems like it's missing some integration somehow... Maybe there's a version of `adapt` that expects it

Maybe my ideal version of your code would be 
```typescript
import { numberAdapter } from '@state-adapt/core/adapters';

interface User {
  name: string;
  age: number;
}

const initialUser: User = {
  name: 'John Doe',
  age: 35,
};

const userStore = adapt(initialUser, [numberAdapter]);
```

What this would do is determine that every property of type `number` inside `initialUser` should get the `numberAdapter` state changes. So automatically you _could_ have 

```typescript
const userStore = adapt({ name: "John Doe", age: 35 }, [numberAdapter]);
userStore.incrementAge();
```

It would still be better to not put the object inline, but anyway. That would be possible. 

Stores take initial state because they aren't abstract. Think of it like calling `new` on an adapter. Adapters are like classes. But we can also define default state in classes, so maybe we can do the same for adapters to allow for automatic "method" generation there too:

```typescript
interface User {
  name: string;
  age: number;
}

const defaultUser: User = {
  name: 'John Doe',
  age: 35,
};

const userAdapter = generateAdapter(defaultUser)(numberAdapter);
```

(idk about `generate`)

If an adapter isn't provided for a property type, it just creates a basic adapter for that type. 

This is starting to feel exciting. This feels like how StateAdapt should have done nested adapters from the beginning. 

One issue though... How does JavaScript determine programmatically what type an adapter is for? I supposed all adapters that follow this pattern would need to have their own `defaultState`. And then it could be a recursive structure that checks all the leaf properties of every type so that if all properties `typeof` are the same, then the parent object types are the same. 

If runtime types are checked this way, it could get expensive every time a new joined adapter is created. So maybe the type could be serialized instead. I think state should be serializable, so this could work. Object keys would need to be sorted. 

```typescript
const aAdapter = createAdapter(0)();
const bAdapter = createAdapter(0)({
  specialBChange: b => b + 99,
});

// Both get a property like `_defaultState: { type: number; value: 0 }`

const abAdapter = generateAdapter({a: 0, b: 1})(aAdapter, bAdapter);

// Oh no... can we prevent multiple adapters defined for the same type? Or should we merge all state changes and selectors defined for the same type from different adapters? 
// idk, let's start over with this example

const aAdapter = createAdapter(0)(); // _defaultState: { type: 'number', value: 0 }
const bAdapter = createAdapter('0')(); // _defaultState: { type: 'string', value: '0' }

const abAdapter = generateAdapter({ a: 0, b: '0' })(aAdapter, bAdapter);
// _defaultState: { type: '{a:number;b:string}', value: { a: 0, b: '0' }}

const adapter = generateAdapter({ ab: { a: 0, b: '0' }})(abAdapter);
```

It knows to use `abAdapter` for the `ab` property because it will serialize the `ab` default state's type and come up with the same type as `abAdapter._defaultState`

---

I need to think about this more. I might be on the verge of something much simpler.

**Comment by mfp22** [comment]
The type serialization thing may not work if it's like `null | number` because the `defaultState` might be one and the adapter another. 

I thought maybe I could just equate types if at least the property names are the same, and be loose like that, because who cares if JS creates extra stuff if TypeScript still limits it? But then I realized this could result in stuff being overridden: 

```typescript
interface User {
  id: string;
}

interface Item {
  id: number;
}

interface State {
  user: User;
  item: Item;
}

const adapter = generateAdapter(defaultState)(userAdapter, itemAdapter)
```

Both with only serialized `_defaultState.type` would be the same, but then which adapter methods take precedence? 

I thought about giving each adapter a token for its type, and it would be like dependency injection—but then realized you would have to map each state property to the token anyway.

Unless... I just kept a global map of all types with adapters. But then there could only be one adapter per type. At least in its implementation. Unless I merged all of _those_ state changes/selectors... then it would be a naming collision again. What are the odds someone creates 2 different adapters for the same type and names a state change function the same thing in both, but then implements it differently in each? I could definitely see that happening. This would probably be too much deviation between JavaScript and TypeScript anyway. 

What if classes are the answer? Classes are a code smell to me because of imperative mutation patterns, but what if they reduce boilerplate by a lot?

```diff
-interface User {
-  name: string;
-  age: number;
-}
-
-const defaultUser: User = {
-  name: 'John Doe',
-  age: 35,
-};
+class User {
+  name = 'John Doe';
+  age = 35;
+}

-const userAdapter = generateAdapter(initialUser)(numberAdapter);
+const userAdapter = generateAdapter(User)(numberAdapter);
```

Now then if there's a parent state:

```typescript
class State {
  user = new User();
}

const adapter = generateAdapter(State)(userAdapter)
```

the `userAdapter` can have its `_defaultState.type = User` and now `generateAdapter` can check each property state for whether it's a primitive and the type matches, or if it's a class and `defaultState[prop] instanceof adapter._defaultState.type`.

In a single adapter generation, we can assume that each property of a type will only have 1 adapter provided for it.

And with the `instanceof` relationship, we can also vary the parent `defaultState`:

```typescript
class A {
  constructor(
    prop = 'prop',
  ) {}
}

const aAdapter = generateAdapter(A)(stringAdapter); // stringAdapter matches typeof A['prop']

class State {
  a = new A('prop override');
}

const store = adapt(new State(), [aAdapter]);
```

So in this pattern, `new` is only called when an actual store is being created. With the `[]` 2nd argument, it knows to match each adapter inside against each property of `new State()` to do an `instanceof` check for the adapter's type. 

---

Does it seem like I'm approaching something good?

**Comment by mfp22** [comment]
Alright, this and a lot of other things you brought up, and a conversation with Nate Abele (creator of UI Router for AngularJS) made me rethink pretty much all of StateAdapt over this weekend. 

Basically, this needs to be StateAdapt 3.0. I rethought ideal syntax, now I will validate, then create the smoothest upgrade path possible. 

Here are some of the changes:

I realized that the adapters can be the source of initial state. No need to even declare it outside. And when creating a store, you can swap out the default state with something else. So rather than having 3 parallel structures: `State`, `initialState`, and `adapter`; they're now all contained in the adapter declaration. Adapters will be callable, which will swap the default state out. 

Here's how your original example would probably look:

```ts
const userStore = adapt({ name: "John Doe", age: counterAdapter(35) }).use();
```

`.use()` is the same as subscribing to an observable. It makes the store active. But with signals, this explicit `use` call is necessary. 

Alternatively, if you aren't reusing state logic yet:

```ts
const userStore = adapt({
  name: 'John Doe',
  age: buildAdapter(35, {
    increment: age => age + 1,
  }),
}).use();
```

I guess you could also do:

```ts

const userStore = adapt({ name: 'John Doe', age: 35 }, {
  incrementAge: state => ({ ...state, age: age + 1 }),
}).use();
```

But the 2nd example would be better for sure.

But look how nice this would be in a full example previously using `joinAdapters`:

```diff
-export interface KeyboardState {
-  up: boolean;
-  down: boolean;
-  left: boolean;
-  right: boolean;
-  pause: boolean;
-  sound: boolean;
-  reset: boolean;
-  drop: boolean;
-  hold: boolean;
-}
-
-const adapter = joinAdapters<KeyboardState>()({
-  up: booleanAdapter,
-  down: booleanAdapter,
-  left: booleanAdapter,
-  right: booleanAdapter,
-  pause: booleanAdapter,
-  sound: booleanAdapter,
-  reset: booleanAdapter,
-  drop: booleanAdapter,
-  hold: booleanAdapter
-})();
-
-export const createInitialState = (): KeyboardState => ({
-  up: false,
-  down: false,
-  left: false,
-  right: false,
-  pause: false,
-  sound: false,
-  reset: false,
-  drop: false,
-  hold: false
-});
-
-export const keyboardStore = adapt(createInitialState(), adapter);
+export const keyboardStore = adapt({
+  up: false,
+  down: false,
+  left: false,
+  right: false,
+  pause: false,
+  sound: false,
+  reset: false,
+  drop: false,
+  hold: false
+}).use();
```

And it's really flexible. You can mix states with adapters:
```ts
export const keyboardStore = adapt({
  up: false,
  down: false,
  left: false,
  right: false,
  pause: false,
  sound: someOtherAdapter(false),
  reset: false,
  drop: false,
  hold: false
}).use();
```

Lots of other changes too. I will have to get some feedback and strategize how to help everyone migrate.

**Comment by Evertt** [comment]
This all looks very cool indeed. just out of curiosity, does this or will there also be any changes made to how sources (and possibly sinks?) work? Did you do some brainstorming about that too?

**Comment by mfp22** [comment]
Pretty much what I've said in this thread.
1. Sources will emit plain values, not actions, and have `type` as a property
2. Sources will be callable directly
3. Sources can be defined with a mapper as a 2nd argument
4. Sources with probably be created now with a `source` function instead of `new Source`

Maybe 1 or 2 of those weren't mentioned, but probably half of them were what you asked for or similar. I still don't want to make them able to be changed without a direct reference though.

**Comment by mfp22** [comment]
Or maybe it was the other thread about your Svelte wishlist

**Comment by Evertt** [comment]
Cool, I look forward to all of that. Sounds like quite major rewrite, isn't it?

**Comment by mfp22** [comment]
Yeah, pretty major. 

Quick question. I'm validating some TypeScript right now. I'm trying to get something like [RxJS's pipe](https://github.com/ReactiveX/rxjs/blob/master/packages/observable/src/observable.ts) but with another possible argument other than `UnaryFunction`. Basically I want adapters themselves to act like `pipe` functions to pass in higher order functions like `withLocalStorage` like `someAdapter(withLocalStorage())` but also so I can pass in a different `defaultState` like `someAdapter(withLocalStorage())(() => newDefaultState)` but TS is having a difficult time inferring potential arguments I might pass into `withLocalStorage` from the adapter. I can get it like RxJS's pipe operator when it's only the adapter enhancers, but not when it could also be a new state. 

Anyway, my temporary workaround is to have

```ts
const userStore = adapt({
  name: "John Doe",
  age: counterAdapter.state(35),
}).use();
```

instead of 

```ts
const userStore = adapt({
  name: "John Doe",
  age: counterAdapter(35),
}).use();
```

Not the end of the world. I know some people would prefer it. What do you think?

**Comment by Evertt** [comment]
I'd love to help, but let me see if I understand your question correctly first.

What I think I understand is that you want the `someInputArg` in `someAdapter(someInputArg)` to have 3 possible types:

1. `someInputArg` could be of type `State`.
2. `someInputArg` could be of type `Factory<State>` which is basically `() => State`.
3. `someInputArg` could be of type `Middleware` which is a higher order function that returns a `UnaryFunction ` or whatever you want to call it. Which basically means `someInputArg` would be of type `(...args: unknown[]) => (inputState: InputState) => OutputState`

So in the latter case, `someInputArg` would be used as `someInputArg(possibleFirstArg, possibleSecondArg)`, just like in RxJS where one might do `scan((total, n) => total + n), 0)`.

Am I on the right track so far? If yes, then the last part isn't totally clear to me. When you say:

> but TS is having a difficult time inferring potential arguments I might pass into withLocalStorage from the adapter

Do you mean TS is having a hard time inferring the types of `possibleFirstArg` and `possibleSecondArg`? Or are you saying that TS is having a hard time inferring the types of `InputState` and `OutputState`?

**Comment by mfp22** [comment]
[I solved it but here's my message with half of it typed before I solved it]

Yes, you got it. 

I guess I could have shared some code. [Playground Link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAyglgcwHYEMA2BnAPDYLgQB8UAvFAEoQDGA9gE4AmWGwdcSCANFABQv4QAXLDwEAlKWIokIQgG4AUKEgUIDAK5UIdbLgHEylWo2at2XXvwLC9BbmBQg0NFA2HSQEksVtFFS8GgAQQYUMAI6HFEIbhgoCAAPAiQGDFhEVEwo-W5yeKSIFLTKDS0dbIJCAygAbwUoKGUhKAByVzCIlsUGhAhgABEIADMUdTRgX2EeLx9o7qg6NU1tDCmZinmMDPRV3nWYTei+G2ioAB9eackRATExYRCO7QqY2Fz5eqgcACZ85NT0sgdi9CLlfol-sUlmVdNEqjxPg0IABGKYoVFQR7hZ6+WLvdZYiIvWLfMGgz73TGhbGRXGwKAAMlgpIojIo3w+DSwQRZQQAzIQeCi0RjCTjoniKIQCSyIN80fLMd8CXzKfzFABffxDdRIKjAOA0JBQKiLARi2lwhENdhwA3oSY3OwKNXUol0mpQKwQPjRE63f0EKAa7g1DXEOoNRbAdR0Y06vUGo1Qdo030CQPQC48K7eJ0QO61RELPqx42piLzDUptIeTUKBQAekbUAA7naABbwIGYBtNQGZSgxuOwgSxP6FAHdzKjyrVSNQADaAGkoOwoABrCAgGhDWAAXWEQ7LABVAjgV-uPlqAipj3Hp8C6TBkb4MNwAHIQVswCBoajAPQaQQpOaSPlkL5vlUpBFg0K7bmuxpbjue5fj+f4AUBh68MW3qZpwxZgHQNBgBgAD87gyARDTrPeSBnpAWBob+-76kB8EgFemr+P24EYAA6p2ACy9AQFgny4hJE5FAOT5wtRUBoQAQs4VAbtJU7bBBWl0bOrwwFUBHztWTJ0XxxK1CZsloLpFkGaGVlxEyymqRuHwKAm+qGsa7bAF2WnYBJEpSSBMnmb45INHkoUAiUyzlBFCkwOCBRhQFOA6aWI52YZCiCp8Ww9rsyUEZS1opm62gPJVlpjm8UqldcFriQ0kkNJ6K6IZu267tZAnCaJdkkoQ2F8YJfkiYsQ3MoQl7Bgp5B5bBJbDsaOYVlVKYyOsG10BIKC1jI9ZNi2BAsA2tBICwFVPHQMGmhA5o1TwAAMYjlb5-lFTwC4NH55jCGkeYYAAdN6CkamICmfXxP3Fv9HCKkDxCgwjXCfJDpVg0cyJvf4u0g4spQrNMigE4VM7TNjAhkzVYMBVTaO07d9PfWIINo98zM0iDvQDMMozjL4pMnVAAACwAYAAtIkkD6jLdDEXdGAdjQYwMFAABG0BIOoAC22t0Aol3XQkgNmBwMEE3zgwjGMExHGIchAA)

Here it's working with `.state(10)` but I'd like it working with just `(10)`. So you can change `state` to be another overload .... Wait, I just tried it and it's working this time. Strange... Maybe `withSignals` needed more explicit typing or something. 

But since both are evidently possible, which do you prefer?

```ts
export const keyboardStore = adapt({
  left: false,
  right: someOtherAdapter(false),
  pause: someOtherAdapter(false),
  sound: someOtherAdapter(false),
});

export const keyboardStore = adapt({
  left: false,
  right: someOtherAdapter.state(false),
  pause: someOtherAdapter.state(false),
  sound: someOtherAdapter.state(false),
});
```

I'm assuming you like the minimal version over the explicit version.

**Comment by mfp22** [comment]
That demo was just using trivially inferable state with no adapters necessary. Now I need to figure out how to infer `State` from something like 
```ts
{
  left: false,
  right: someOtherAdapter(false),
  pause: someOtherAdapter(false),
  sound: someOtherAdapter(false),
}
```

**Comment by mfp22** [comment]
Oh, I found the problem. It's not even matching against the first overload in my example. If you erase the other `with` functions so there's only 1, you see the problem: https://www.typescriptlang.org/play?#code/C4TwDgpgBAyglgcwHYEMA2BnAPDYLgQB8UAvFAEoQDGA9gE4AmWGwdcSCANFABQv4QAXLDwEAlKWIokIQgG4AUKEgUIDAK5UIdbLgHEylWo2at2XXvwLC9BbmBQg0NFA2HSQEksVtFFS8GgAQQYUMAI6HFEIbhgoCAAPAiQGDFhEVEwo-W5yeKSIFLTKDS0dbIJCAygAbwUoKGUhKAByVzCIlsUGhAhgABEIADMUdTRgX2EeLx9o7qg6NU1tDCmZinmMDPRV3nWYeb4baKgAH15pyREBMTFhEI7tCpjYXPl6qBwAJnzk1PTkDtnoRcj9En9iksyrpolUeB8GhAAIxTFAoqAPcJPXyxN7rTERZ6xL6gkEfO4Y0JYyI42BQABksBJFAZFC+7waWCCzKCAGZCDxkaj0QTsdFcRRCPjmRAvqi5RivvjeRS+YoAL7+IbqJBUYBwGhIKBURYCUU02HwhrsOD69CTa52BSqqmE2k1KBWCB8aLHG5+ghQdXcGrq4h1BqLYDqOhG7W6-WGqDtak+gQB6DnHiXbyOiC3WoIhZ9GNGlMRebq5NpDwahQKAD0DagAHdbQALeCAzD1poAzKUaOxmECWK-Qr-LuZEeVaoRqAAbQA0lB2FAANYQEA0IawAC6wkHpYAKoEcMu9+9NQEVEfY1OgbSYEjfBhuAA5CAtmAQNDUYD0Gk4ITmkD5ZM+r5VKQhYNMuW6rkam7brun7fr+-6AQevBFl6GacEWYB0DQYAYAA-O4Mj4Q06x3kgp6QFgqE-n+eqAXBICXhq-h9mBGAAOodgAsvQEBYB8OLieORT9o+sJUVAqEAELOFQ65SZO2zgZptEzi8MBVPhc5VoytG8UStTGTJaA6eZ+khpZcSMkpKnru8CjxnqBpGm2wCdpp2DieKknAdJZm+GSDR5CF-wlMs5ThfJMBggUoX+Tg2klsOtkGQoAofFs3a7El+EUlayauto9wVRao6vJKJVXOaYkNBJDQesuCEbluO5WfxQkibZxKEFhvECb5wmLINTKEBeQbyeQuUwcWQ5Gtm5aVcmMjrOtdASCgNYyHWjbNgQLD1rQSAsOVjx0NBJoQGa1U8AADGIZU+X5hU8PODS+eYwhpLmGAAHRevJ6piCV2brEir3+DtwOLKUKzTIoCMFdO0yg3MCjo-5WN-RwaPVcDvQDMMozjL4qPHVAAACwAYAAtIkkB6izdBEbdGDtjQYwMFAABG0BIOoAC2wt0AoF1XQkANmBw0EI2TgwjGMEzRKjQA

**Comment by mfp22** [comment]
I still haven't figured that out, but I figured out how to automatically create joined adapters from states, and it's recursive. So for example

```ts
const adapter = buildAdapter({
  settings: {
    theme: {
      stuff: false,
    },
  },
});
```

will generate this state change function: `store.setSettingsThemeStuff(true)`

Now I'm going to make it so it defines some bare minimum reducers for each primitive type (boolean, number, etc)
But it's super easy to substitute in a different adapter:

```ts
const adapter = buildAdapter({
  settings: {
    theme: {
      stuff: booleanDeluxeAdapter(false),
    },
  },
});
```

I'm loving this DX so far. 

I personally prefer being able to pass in default state like that rather than in `.state`. If I can't get the overrides to play okay with each other, I think I might rather have `buildAdapter(0).add(withSignals(...))(() => newDefaultState)` than `buildAdapter(0)(withSignals(...)).state(() => newDefaultState)`

Replacing state will probably be super common.

**Comment by Evertt** [comment]
@mfp22 just checking, are you still asking for help? Or are you now mainly using this thread as a soundboard? :-)

I'd be happy to help wherever I can, but in that case I would need to see another repro in (TS playground)[typescriptlang.org/play] to see how you're building it so far and where the exact problem occurs.

**Comment by mfp22** [comment]
Yeah this thread got confusing. I figured it out, but only after I wrote all these messages. The "Edit:" comment was the last thing I wrote.

Today I added a function that allows passing the type directly in. Because to get all these types to work, I needed a generic other than State, so you can't pass that in directly. So for people who already have types defined and don't want the object literal to show up in all their type definitions, it would look like this

```ts
const adapter = buildAdapter({
  settings: {
    theme: {
      stuff: booleanDeluxeAdapter(false),
    },
  },
})(withType<AppState>());
```

Not amazing, but this will probably be relatively uncommon.

**Comment by mfp22** [comment]
I guess I'm using this as a sounding board, but I'm interested in your opinions if you have any.

**Comment by Evertt** [comment]
> I guess I'm using this as a sounding board, but I'm interested in your opinions if you have any.

I don't have any for now, I also like the direction you're going in. I might form some opinions / preferences once I actually start using the new way of building adapters, but even then I'm not 100% sure how valuable my opinions will be. Since our recent discussion around how `Source` should work, especially the part about my special `pipe` method, has made me realize that I probably still don't understand declarative reactive programming enough to be able to distinguish between good ideas and bad ideas.

So for now, I'm happy to just sit back and watch what you'll create. And I'm also happy to help you get unstuck whenever you're getting stuck with TS. And then my main intention is to try to get used to using your library the way you intended until I fully understand every design decision that you made. :-)

**Comment by mfp22** [comment]
I'm going to close this issue and refer to it from a new issue focused on new StateAdapt 5 syntax. 

However, this is more of an optimization that will come after looking into or refining the Spaghetti ESLint plugin and probably adding a lot of of wrapper components and APIs for enabling declarative code.
